File: project/android/apps/MnnLlmApp/app/src/main/java/com/alibaba/mls/api/HfApiClient.java
// Created by ruoyi.sjd on 2024/12/25.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.

package com.alibaba.mls.api;

import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.*;
import retrofit2.converter.gson.GsonConverterFactory;

public class HfApiClient {

    private static final String TAG = "HfApiClient";

    public static final String HOST_DEFAULT = "huggingface.co";
    public static final String HOST_MIRROR = "hf-mirror.com";

    private String host;
    private HfApiService apiService;
    private OkHttpClient okHttpClient;
    private static HfApiClient sBestClient;

    public static void setBestClient(HfApiClient apiClient) {
        sBestClient = apiClient;
    }

    public static HfApiClient getBestClient() {
        if (sBestClient != null) {
            return sBestClient;
        }
        return null;
    }

    public HfApiClient(String defaultHost) {
        this.host = defaultHost;
        // Initialize Retrofit
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://" + host)
                .addConverterFactory(GsonConverterFactory.create())
                .client(createOkHttpClient())
                .build();
        apiService = retrofit.create(HfApiService.class);
    }

    public okhttp3.OkHttpClient getOkHttpClient() {
        return okHttpClient;
    }

    private okhttp3.OkHttpClient createOkHttpClient() {
        okhttp3.OkHttpClient.Builder builder = new okhttp3.OkHttpClient.Builder();
        builder.connectTimeout(30, TimeUnit.SECONDS);
        builder.readTimeout(30, TimeUnit.SECONDS);
        okHttpClient =  builder.build();
        return okHttpClient;
    }

    public String getHost() {
        return this.host;
    }

    // Searches repositories based on a keyword
    public void searchRepos(String keyword, RepoSearchCallback callback) {
        Call<List<HfRepoItem>> call = apiService.searchRepos(keyword, "taobao-mnn", 500);
        call.enqueue(new Callback<List<HfRepoItem>>() {
            @Override
            public void onResponse(Call<List<HfRepoItem>> call, Response<List<HfRepoItem>> response) {
                callback.onSuccess(response.body());
            }

            @Override
            public void onFailure(Call<List<HfRepoItem>> call, Throwable t) {
                callback.onFailure(t.getMessage());
            }
        });
    }

    // Retrieves repository information
    public void getRepoInfo(String repoName, String revision, RepoInfoCallback callback) {
        Call<HfRepoInfo> call = apiService.getRepoInfo(repoName, revision);
        call.enqueue(new Callback<HfRepoInfo>() {
            @Override
            public void onResponse(Call<HfRepoInfo> call, Response<HfRepoInfo> response) {
                callback.onSuccess(response.body());
            }

            @Override
            public void onFailure(Call<HfRepoInfo> call, Throwable t) {
                callback.onFailure(t.getMessage());
            }
        });
    }

    // Callback interfaces
    public interface RepoSearchCallback extends CallbackWrapper<List<HfRepoItem>> {
        void onSuccess(List<HfRepoItem> hfRepoItems);
        void onFailure(String error);
    }

    public interface RepoInfoCallback extends CallbackWrapper<HfRepoInfo> {
        void onSuccess(HfRepoInfo hfRepoInfo);
        void onFailure(String error);
    }

    // Wrapper interface to generalize callbacks
    private interface CallbackWrapper<T> {
        void onSuccess(T result);
        void onFailure(String error);
    }

}

File: apps/Android/Mnn3dAvatar/app/src/main/java/com/alibaba/mls/api/HfApiClient.kt
// Created by ruoyi.sjd on 2024/12/25.
// Copyright (c) 2024 Alibaba Group Holding Limited All rights reserved.
package com.alibaba.mls.api

import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

class HfApiClient(@JvmField val host: String) {
    private val apiService: HfApiService
    var okHttpClient: OkHttpClient? = null
        private set

    init {
        // Initialize Retrofit
        val retrofit = Retrofit.Builder()
            .baseUrl("https://$host")
            .addConverterFactory(GsonConverterFactory.create())
            .client(createOkHttpClient()!!)
            .build()
        apiService = retrofit.create(HfApiService::class.java)
    }

    private fun createOkHttpClient(): OkHttpClient? {
        val builder: OkHttpClient.Builder = OkHttpClient.Builder()
        builder.connectTimeout(30, TimeUnit.SECONDS)
        builder.readTimeout(30, TimeUnit.SECONDS)
//        val okHttpClient:OkHttpClient = builder.build()
        okHttpClient = builder.build()
        return okHttpClient
    }

    // Searches repositories based on a keyword
    fun searchRepos(keyword: String?, callback: RepoSearchCallback) {
        val call = apiService.searchRepos(keyword, "taobao-mnn", 500, "downloads")
        call.enqueue(object : Callback<List<ModelItem>> {
            override fun onResponse(
                call: Call<List<ModelItem>>,
                response: Response<List<ModelItem>>
            ) {
                callback.onSuccess(response.body()!!)
            }

            override fun onFailure(call: Call<List<ModelItem>>, t: Throwable) {
                callback.onFailure(t.message)
            }
        })
    }

    // Retrieves repository information
    fun getRepoInfo(repoName: String?, revision: String?, callback: RepoInfoCallback) {
        val call = apiService.getRepoInfo(repoName, revision)
        call!!.enqueue(object : Callback<HfRepoInfo?> {
            override fun onResponse(call: Call<HfRepoInfo?>, response: Response<HfRepoInfo?>) {
                callback.onSuccess(response.body())
            }

            override fun onFailure(call: Call<HfRepoInfo?>, t: Throwable) {
                callback.onFailure(t.message)
            }
        })
    }

    // Callback interfaces
    interface RepoSearchCallback :
        CallbackWrapper<List<ModelItem>> {
        override fun onSuccess(hfModelItems: List<ModelItem>)
        override fun onFailure(error: String?)
    }

    interface RepoInfoCallback : CallbackWrapper<HfRepoInfo?> {
        override fun onSuccess(hfRepoInfo: HfRepoInfo?)
        override fun onFailure(error: String?)
    }

    // Wrapper interface to generalize callbacks
    interface CallbackWrapper<T> {
        fun onSuccess(result: T)
        fun onFailure(error: String?)
    }

    companion object {
        private const val TAG = "HfApiClient"

        const val HOST_DEFAULT: String = "huggingface.co"
        const val HOST_MIRROR: String = "hf-mirror.com"

        private var sBestClient: HfApiClient? = null

        @JvmStatic
        var bestClient: HfApiClient?
            get() {
                if (sBestClient != null) {
                    return sBestClient
                }
                return null
            }
            set(apiClient) {
                sBestClient = apiClient
            }
    }
}

